File: 076-minimum-window-substring/minimum-window-substring.py
# -*- coding:utf-8 -*-


# Given a string S and a string T, find the minimum window in S which will contain all the characters in T in complexity O(n).
#
#
#
# For example,
# S = "ADOBECODEBANC"
# T = "ABC"
#
#
# Minimum window is "BANC".
#
#
#
# Note:
# If there is no such window in S that covers all characters in T, return the empty string "".
#
#
# If there are multiple such windows, you are guaranteed that there will always be only one unique minimum window in S.


class Solution(object):
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        d = dict()
        for c in t:
            if c in d:
                d[c] += 1
            else:
                d[c] = 1
        counter = len(t)
        begin = 0
        end = 1
        res = float('inf')
        res_str = ""
        while end <= len(s):
            c = s[end - 1]
            if c in d:
                d[c] -= 1
                if d[c] >= 0:
                    counter -= 1
            while counter == 0 and begin < end:
                if (end - begin) < res:
                    res = end - begin
                    res_str = s[begin:end]
                dc = s[begin]
                if dc in d:
                    d[dc] += 1
                    if d[dc] > 0:
                        counter += 1
                begin += 1
            end += 1
        return res_str
        

File: 076-minimum-window-substring/minimum-window-substring.cpp
// Given a string S and a string T, find the minimum window in S which will contain all the characters in T in complexity O(n).
//
// Example:
//
//
// Input: S = "ADOBECODEBANC", T = "ABC"
// Output: "BANC"
//
//
// Note:
//
//
// 	If there is no such window in S that covers all characters in T, return the empty string "".
// 	If there is such window, you are guaranteed that there will always be only one unique minimum window in S.
//
//


class Solution {
public:
    string minWindow(string s, string t) {
        int ns = s.size();
        int nt = t.size();
        unordered_map<char, int> m;
        for (int i = 0; i < nt; i++) {
            char c = t[i];
            if (m.find(c) == m.end())
                m[c] = 0;
            m[c] += 1;
        }
        int begin = 0, end = 0;
        int wlen = ns + 1;
        string window = "";
        while (end < ns) {
            char c = s[end];
            if (m.find(c) != m.end()) {
                m[c] -= 1;
                if (m[c] >= 0) nt -= 1;
            }
            while (nt == 0 && begin <= end) {
                int l = end - begin + 1;
                if (l < wlen) {
                    wlen = l;
                    window = s.substr(begin, l);
                }
                char b = s[begin];
                if (m.find(b) != m.end()) {
                    m[b] += 1;
                    if (m[b] > 0) nt += 1;
                }
                begin += 1;
            }
            end += 1;
        }
        return window;
    }
};
